# Titanic Dataset: Understanding, Cleaning and Preparation

## First, we need our libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Uploading and Understanding the Dataset

In [2]:
df_titanic = pd.read_csv('Titanic.tsv', sep='\t')

### Examining the columns and rows

In [3]:
df_titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'ship'],
      dtype='object')

In [4]:
df_titanic.shape

(894, 13)

In [5]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 894 entries, 0 to 893
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  894 non-null    int64  
 1   Survived     894 non-null    int64  
 2   Pclass       894 non-null    float64
 3   Name         894 non-null    object 
 4   Sex          894 non-null    object 
 5   Age          721 non-null    object 
 6   SibSp        894 non-null    int64  
 7   Parch        894 non-null    object 
 8   Ticket       893 non-null    object 
 9   Fare         893 non-null    object 
 10  Cabin        208 non-null    object 
 11  Embarked     892 non-null    object 
 12  ship         894 non-null    object 
dtypes: float64(1), int64(3), object(9)
memory usage: 90.9+ KB


#### As can be seen from above we have some null values, So I wanted to see how many null values we have in each column

In [6]:
df_titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            173
SibSp            0
Parch            0
Ticket           1
Fare             1
Cabin          686
Embarked         2
ship             0
dtype: int64

#### We have so many null values especially in 'Cabin' column but also we have approx %15 of null values in 'Age' column.

#### Trying to see some of the values by using 'head()' or 'tail()' functions. 

In [7]:
df_titanic.head(10)

PassengerId  Survived  Pclass  \
0            1         0     3.0   
1            2         1     1.0   
2            3         1     3.0   
3            4         1     1.0   
4            5         0     3.0   
5            6         0     3.0   
6            7         0     1.0   
7            8         0     3.0   
8            9         1     3.0   
9           10         1     2.0   

                                                Name     Sex  Age  SibSp  \
0                            Braund, Mr. Owen Harris    male   22      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38      1   
2                             Heikkinen, Miss. Laina  female   26      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35      1   
4                           Allen, Mr. William Henry    male   35      0   
5                                   Moran, Mr. James    male  NaN      0   
6                            McCarthy, Mr. Timothy J    male   54      0   
7                     Palsson, Master. Gosta Leonard    male    2      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   27      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female   14      1   

  Parch            Ticket     Fare Cabin Embarked     ship  
0     0         A/5 21171     7,25   NaN        S  Titanic  
1     0          PC 17599  71,2833   C85        C  Titanic  
2     0  STON/O2. 3101282    7,925   NaN        S  Titanic  
3     0            113803     53,1  C123        S  Titanic  
4     0            373450     8,05   NaN        S  Titanic  
5     0            330877   8,4583   NaN        Q  Titanic  
6     0             17463  51,8625   E46        S  Titanic  
7     1            349909   21,075   NaN        S  Titanic  
8     2            347742  11,1333   NaN        S  Titanic  
9     0            237736  30,0708   NaN        C  Titanic

#### We have some problems with 'Pclass', 'Name', 'Fare' columns can be seen from here but I will check them step by step in the future cells.

#### I also wanted to check if we have duplicate values.

In [8]:
df_titanic.duplicated().sum()

3

#### We have 3 duplicated values. But maybe we have other duplicate values that shouldn't be available in the columns. Exp. 'PassengerId' column, maybe 'Ticket' column, too.

## Looking at each columns and understanding them in detail. And Fixing them step by step.

### Examining each column step by step to find problems.

### PassengerId

In [9]:
df_titanic['PassengerId'].duplicated().sum()

4

#### As I mentioned above we have some duplicate values in PassengerId column but shouldn't be.

#### We saw that we don't have any null values in this column.

#### But maybe we have some random or improper values.

In [10]:
df_titanic['PassengerId'].value_counts()

 225    3
 11     3
 1      1
 603    1
 592    1
       ..
 304    1
 305    1
 306    1
 307    1
-12     1
Name: PassengerId, Length: 890, dtype: int64

In [11]:
for x in df_titanic['PassengerId']:
  if x < 0:
    print(x)

-12


#### We have only 1 minus value, now I will check for other improper values.

In [12]:
for x in df_titanic['PassengerId']:
    if str(x).isalpha() == True:
        print(x)
    

#### As can be seen from above we have only numbers but we have 1 minus value and we have some duplicated values. So I think we can delete this column and re-create it in order to have a correct 'PassengerId' column.

In [13]:
df_titanic.drop(['PassengerId'], axis=1, inplace=True)

In [14]:
df_titanic.insert(0, "PassengerId", range(1,895))

In [15]:
df_titanic.head(5)

PassengerId  Survived  Pclass  \
0            1         0     3.0   
1            2         1     1.0   
2            3         1     3.0   
3            4         1     1.0   
4            5         0     3.0   

                                                Name     Sex Age  SibSp Parch  \
0                            Braund, Mr. Owen Harris    male  22      1     0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1     0   
2                             Heikkinen, Miss. Laina  female  26      0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1     0   
4                           Allen, Mr. William Henry    male  35      0     0   

             Ticket     Fare Cabin Embarked     ship  
0         A/5 21171     7,25   NaN        S  Titanic  
1          PC 17599  71,2833   C85        C  Titanic  
2  STON/O2. 3101282    7,925   NaN        S  Titanic  
3            113803     53,1  C123        S  Titanic  
4            373450     8,05   NaN        S  Titanic

#### So now we have a correct 'PassengerId' column

### Survived

#### Now I will check for the survived column with improper or random values. We don't have null values in this column, too. And also we can have duplicate values as you can imagine.

In [16]:
df_titanic['Survived'].value_counts()

 0    548
 1    345
-4      1
Name: Survived, dtype: int64

In [17]:
print(548+345+1)

894


#### We don't have other wrong value than '-4' in this column so we only need to fix it. I think this value might be 0 becuse it is closer to 0. But also it can be all wrong in this row I should check it first.

In [18]:
df_titanic.loc[df_titanic['Survived'] == -4]

PassengerId  Survived  Pclass                                   Name  \
858          859        -4     3.0  Baclini, Mrs. Solomon (Latifa Qurban)   

        Sex Age  SibSp Parch Ticket     Fare Cabin Embarked     ship  
858  female  24      0     3   2666  19,2583   NaN        C  Titanic

#### It seems that only Survived column is wrong so I will fix it with 0.

In [19]:
df_titanic['Survived'] = df_titanic['Survived'].replace(-4, 0)

In [20]:
df_titanic['Survived'].value_counts()

0    549
1    345
Name: Survived, dtype: int64

#### So we have fixed the 'Survived' column, too.

### Pclass

#### As I mentioned above we have some problems with Pclass that can bee seen in the first eye.

In [21]:
df_titanic['Pclass'].head(5)

0    3.0
1    1.0
2    3.0
3    1.0
4    3.0
Name: Pclass, dtype: float64

#### ıt shouldn't be float, we can change it to integer or string like 'first', 'second' and 'third' classes. I think I will change them to strings but first we should check for the improper and random values. We don't have null values as we looked above.

In [22]:
df_titanic['Pclass'].value_counts()

 3.0    493
 1.0    216
 2.0    184
-2.0      1
Name: Pclass, dtype: int64

In [23]:
df_titanic['Pclass'].value_counts().sum()

894

#### We only have on wrong value which is '-2.0'. I think it should be '2.0' maybe put minus sign by mistake.

In [24]:
df_titanic['Pclass'] = df_titanic['Pclass'].replace(-2.0, 2.0)

In [25]:
df_titanic['Pclass'].value_counts()

3.0    493
1.0    216
2.0    185
Name: Pclass, dtype: int64

#### Now I will change the values into strings.

In [26]:
df_titanic['Pclass'] = df_titanic['Pclass'].replace([3.0, 2.0, 1.0], ['Third', 'Second', 'First'])

In [27]:
df_titanic['Pclass'].value_counts()

Third     493
First     216
Second    185
Name: Pclass, dtype: int64

In [28]:
df_titanic['Pclass'].head(5)

0    Third
1    First
2    Third
3    First
4    Third
Name: Pclass, dtype: object

#### So we have fixed this column, too.

### Name

In [29]:
df_titanic['Name'].head(5)

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: object

#### At first, I think we should seperate the 'Name' into 3 different columns: 'Title', 'FirstName', 'LastName'.

#### But we may have some problems with the seperating letters like ','. But in order to see it I think we should seperate the columns first.

In [30]:
df_titanic['FirstName'] = df_titanic['Name'].str.split(',').str[1]

In [31]:
df_titanic['LastName'] = df_titanic['Name'].str.split(',').str[0]

In [32]:
df_titanic.head(5)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   

                                                Name     Sex Age  SibSp Parch  \
0                            Braund, Mr. Owen Harris    male  22      1     0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1     0   
2                             Heikkinen, Miss. Laina  female  26      0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1     0   
4                           Allen, Mr. William Henry    male  35      0     0   

             Ticket     Fare Cabin Embarked     ship  \
0         A/5 21171     7,25   NaN        S  Titanic   
1          PC 17599  71,2833   C85        C  Titanic   
2  STON/O2. 3101282    7,925   NaN        S  Titanic   
3            113803     53,1  C123        S  Titanic   
4            373450     8,05   NaN        S  Titanic   

                                     FirstName   LastName  
0                              Mr. Owen Harris     Braund  
1   Mrs. John Bradley (Florence Briggs Thayer)    Cumings  
2                                  Miss. Laina  Heikkinen  
3           Mrs. Jacques Heath (Lily May Peel)   Futrelle  
4                            Mr. William Henry      Allen

In [33]:
df_titanic['Title'] = df_titanic['FirstName'].str.split('.').str[0]

In [34]:
df_titanic['FirstName'] = df_titanic['FirstName'].str.split('.').str[1]

In [35]:
df_titanic.head(5)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   

                                                Name     Sex Age  SibSp Parch  \
0                            Braund, Mr. Owen Harris    male  22      1     0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1     0   
2                             Heikkinen, Miss. Laina  female  26      0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1     0   
4                           Allen, Mr. William Henry    male  35      0     0   

             Ticket     Fare Cabin Embarked     ship  \
0         A/5 21171     7,25   NaN        S  Titanic   
1          PC 17599  71,2833   C85        C  Titanic   
2  STON/O2. 3101282    7,925   NaN        S  Titanic   
3            113803     53,1  C123        S  Titanic   
4            373450     8,05   NaN        S  Titanic   

                                FirstName   LastName  Title  
0                             Owen Harris     Braund     Mr  
1   John Bradley (Florence Briggs Thayer)    Cumings    Mrs  
2                                   Laina  Heikkinen   Miss  
3           Jacques Heath (Lily May Peel)   Futrelle    Mrs  
4                           William Henry      Allen     Mr

#### Now, I will check each columns.

### Title

In [36]:
df_titanic['Title'].isnull().sum()

1

In [37]:
df_titanic['Title'].value_counts()

 Mr              517
 Miss            185
 Mrs             124
 Master           40
 Dr                7
 Rev               6
 Mlle              2
 Major             2
 Col               2
 the Countess      1
 Capt              1
 Ms                1
 Sir               1
 Lady              1
 Mme               1
 Don               1
 Jonkheer          1
Name: Title, dtype: int64

In [38]:
df_titanic['Title'].value_counts().sum()

893

#### So as can be seen from above, we have 1 null value, few improper values that can be fixed but I think in order to do that we have to group the titles in to less number of groups.

#### First, I will look at the null value.

In [39]:
df_titanic.loc[df_titanic['Title'].isnull() == True]

PassengerId  Survived Pclass                         Name   Sex Age  \
891          892         1  First  Mr. Frederick Maxfield Hoyt  male  38   

     SibSp Parch Ticket Fare Cabin Embarked     ship FirstName  \
891      1     0  19943   90   C93        S  Titanic       NaN   

                        LastName Title  
891  Mr. Frederick Maxfield Hoyt   NaN

#### I think this is happening because of seperating problem.

In [40]:
df_titanic['Title'] = df_titanic['Title'].fillna(' Mr')

In [41]:
df_titanic['Title'].isnull().sum()

0

In [42]:
df_titanic['Title'].value_counts()

 Mr              518
 Miss            185
 Mrs             124
 Master           40
 Dr                7
 Rev               6
 Mlle              2
 Major             2
 Col               2
 the Countess      1
 Capt              1
 Ms                1
 Sir               1
 Lady              1
 Mme               1
 Don               1
 Jonkheer          1
Name: Title, dtype: int64

#### I have fixed the Title column. Filled the NaN with the Mr because I can see that it should be Mr

In [43]:
df_titanic['FirstName'].isnull().sum()

1

In [44]:
df_titanic['FirstName'] = df_titanic['FirstName'].fillna(' Frederick Maxfield')

In [45]:
df_titanic['FirstName'].isnull().sum()

0

In [46]:
df_titanic.loc[df_titanic['FirstName'] == ' Frederick Maxfield']

PassengerId  Survived Pclass                           Name   Sex Age  \
224          225         1  First   Hoyt, Mr. Frederick Maxfield  male  38   
520          521         1  First   Hoyt, Mr. Frederick Maxfield  male  38   
678          679         1  First  Hoytt, Mr. Frederick Maxfield  male  38   
891          892         1  First    Mr. Frederick Maxfield Hoyt  male  38   

     SibSp Parch Ticket Fare Cabin Embarked     ship            FirstName  \
224      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   
520      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   
678      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   
891      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   

                        LastName Title  
224                         Hoyt    Mr  
520                         Hoyt    Mr  
678                        Hoytt    Mr  
891  Mr. Frederick Maxfield Hoyt    Mr

#### Now, I have fixed the FirstName column and lastly I will fix the LastName column.

In [47]:
df_titanic['LastName'] = df_titanic['LastName'].replace('Mr. Frederick Maxfield Hoyt', ' Hoyt')

In [48]:
df_titanic.loc[df_titanic['FirstName'] == ' Frederick Maxfield']

PassengerId  Survived Pclass                           Name   Sex Age  \
224          225         1  First   Hoyt, Mr. Frederick Maxfield  male  38   
520          521         1  First   Hoyt, Mr. Frederick Maxfield  male  38   
678          679         1  First  Hoytt, Mr. Frederick Maxfield  male  38   
891          892         1  First    Mr. Frederick Maxfield Hoyt  male  38   

     SibSp Parch Ticket Fare Cabin Embarked     ship            FirstName  \
224      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   
520      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   
678      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   
891      1     0  19943   90   C93        S  Titanic   Frederick Maxfield   

    LastName Title  
224     Hoyt    Mr  
520     Hoyt    Mr  
678    Hoytt    Mr  
891     Hoyt    Mr

#### Now I have fixed this problem and fixed the null Title value. Now continue to check and fix for the Title column again.

In [49]:
df_titanic['Title'].value_counts().sum()

894

In [50]:
df_titanic['Title'].value_counts()

 Mr              518
 Miss            185
 Mrs             124
 Master           40
 Dr                7
 Rev               6
 Mlle              2
 Major             2
 Col               2
 the Countess      1
 Capt              1
 Ms                1
 Sir               1
 Lady              1
 Mme               1
 Don               1
 Jonkheer          1
Name: Title, dtype: int64

#### We have to decrease the number of Titles.

#### Now I will look at FirstName and LastName

### FirstName

In [51]:
df_titanic['FirstName'].value_counts().sum()

894

In [52]:
df_titanic['FirstName'].isnull().sum()

0

#### There is no null values maybe there was 1 before but I have fixed it in the Title section

#### now I will check for the improper and random values.

In [53]:
df_titanic['FirstName'].dtype

dtype('O')

In [54]:
def nonAlphabets(string: str) -> str:
    res = ""
    for char in string:
        if not char.isalpha():
            res += char
    return res

In [55]:
for x in df_titanic['FirstName']:
    print(x, nonAlphabets(x))

 Owen Harris   
 John Bradley (Florence Briggs Thayer)    (  )
 Laina  
 Jacques Heath (Lily May Peel)    (  )
 William Henry   
 James  
 Timothy J   
 Gosta Leonard   
 Oscar W (Elisabeth Vilhelmina Berg)    (  )
 Nicholas (Adele Achem)   ( )
 Marguerite Ru&5$$   &5$$
 Elizabeth  
 William Henry   
 Marguerite Ru&5$$   &5$$
 Hulda Amanda Adolfina    
 (Mary D Kingcome)  (  )
 Eugene  
 Charles Eugene   
 Julius (Emelia Maria Vandemoortele)   (  )
 Fatima  
 Joseph J   
 Lawrence  
 Anna "Annie"   ""
 Marguerite Ru&5$$   &5$$
 Torborg Danira   
 Carl Oscar (Selma Augusta Emilia Johansson)    (   )
 Farred Chehab   
 Charles Alexander   
 Ellen "Nellie"   ""
 Lalio  
 Manuel E   
 William Augustus (Marie Eugenie)    ( )
 Mary Agatha   
 Edward H   
 Edgar Joseph   
 Alexander Oskar   
 Hanna  
 Ernest Charles   
 Augusta Maria   
 Jamila  
 Johan (Johanna Persdotter Larsson)   (  )
 William John Robert (Dorothy Ann Wonnacott)     (  )
 Theodor  
 Simonne Marie Anne Andree     
 Margare

#### I think there are many duplicate values in the dataframe but we cannot find them by using dataframe itself because I have fixed the PassengerId column so we have to find them.

In [56]:
df_titanic.loc[df_titanic['FirstName'] == ' Marguerite Ru&5$$']

PassengerId  Survived Pclass                                Name     Sex  \
10            11         1  Third  Sandstrom, Miss. Marguerite Ru&5$$  female   
13            14         1  Third  Sandstrom, Miss. Marguerite Ru&5$$  female   
23            24         1  Third  Sandstrom, Miss. Marguerite Ru&5$$  female   
892          893         1  Third  Sandstrom, Miss. Marguerite Ru&5$$  female   

    Age  SibSp Parch   Ticket  Fare Cabin Embarked     ship  \
10    4      1     1  PP 9549  16,7    G6        S  Titanic   
13    4      1     1  PP 9549  16,7    G6        S  Titanic   
23    4      1     1  PP 9549  16,7    G6        S  Titanic   
892   4      1     1  PP 9549  16,7    G6        S  Titanic   

              FirstName   LastName  Title  
10    Marguerite Ru&5$$  Sandstrom   Miss  
13    Marguerite Ru&5$$  Sandstrom   Miss  
23    Marguerite Ru&5$$  Sandstrom   Miss  
892   Marguerite Ru&5$$  Sandstrom   Miss

In [57]:
df_titanic.duplicated().sum()

0

#### Finding duplicated values.

In [58]:
df_titanic.duplicated(subset=['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'ship', 'FirstName', 'LastName', 'Title', ]).sum()

4

In [59]:
df_titanic.drop_duplicates(subset=['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'ship', 'FirstName', 'LastName', 'Title', ], keep='first', inplace=True)

In [60]:
df_titanic.loc[df_titanic['FirstName'] == ' Marguerite Ru&5$$']

PassengerId  Survived Pclass                                Name     Sex  \
10           11         1  Third  Sandstrom, Miss. Marguerite Ru&5$$  female   

   Age  SibSp Parch   Ticket  Fare Cabin Embarked     ship  \
10   4      1     1  PP 9549  16,7    G6        S  Titanic   

             FirstName   LastName  Title  
10   Marguerite Ru&5$$  Sandstrom   Miss

In [61]:
df_titanic.shape

(890, 16)

#### I think I have removed duplicate values. Now, we can continue to work with FirstName column.

In [62]:
for x in df_titanic['FirstName']:
    print(x, nonAlphabets(x))

 Owen Harris   
 John Bradley (Florence Briggs Thayer)    (  )
 Laina  
 Jacques Heath (Lily May Peel)    (  )
 William Henry   
 James  
 Timothy J   
 Gosta Leonard   
 Oscar W (Elisabeth Vilhelmina Berg)    (  )
 Nicholas (Adele Achem)   ( )
 Marguerite Ru&5$$   &5$$
 Elizabeth  
 William Henry   
 Hulda Amanda Adolfina    
 (Mary D Kingcome)  (  )
 Eugene  
 Charles Eugene   
 Julius (Emelia Maria Vandemoortele)   (  )
 Fatima  
 Joseph J   
 Lawrence  
 Anna "Annie"   ""
 Torborg Danira   
 Carl Oscar (Selma Augusta Emilia Johansson)    (   )
 Farred Chehab   
 Charles Alexander   
 Ellen "Nellie"   ""
 Lalio  
 Manuel E   
 William Augustus (Marie Eugenie)    ( )
 Mary Agatha   
 Edward H   
 Edgar Joseph   
 Alexander Oskar   
 Hanna  
 Ernest Charles   
 Augusta Maria   
 Jamila  
 Johan (Johanna Persdotter Larsson)   (  )
 William John Robert (Dorothy Ann Wonnacott)     (  )
 Theodor  
 Simonne Marie Anne Andree     
 Margaret Delia   
 William John   
 Denis  
 Bridget  
 You

#### Now, Going to find and fix improper values

In [63]:
special_characters = "!@#$%^&*-+?_=,<>/123456789"

In [64]:
for x in df_titanic['FirstName']:
    for c in special_characters:
        if c in x:
            print(x)

 Marguerite Ru&5$$
 Marguerite Ru&5$$
 Marguerite Ru&5$$
 Carl/Charles Peter
 of (Lucy Noel Martha Dyer-Edwards)


In [65]:
df_titanic.loc[df_titanic['FirstName'] == ' Marguerite Ru&5$$']

PassengerId  Survived Pclass                                Name     Sex  \
10           11         1  Third  Sandstrom, Miss. Marguerite Ru&5$$  female   

   Age  SibSp Parch   Ticket  Fare Cabin Embarked     ship  \
10   4      1     1  PP 9549  16,7    G6        S  Titanic   

             FirstName   LastName  Title  
10   Marguerite Ru&5$$  Sandstrom   Miss

In [66]:
df_titanic['FirstName'] = df_titanic['FirstName'].replace(' Marguerite Ru&5$$', ' Marguerite')

In [67]:
for x in df_titanic['FirstName']:
    for c in special_characters:
        if c in x:
            print(x)

 Carl/Charles Peter
 of (Lucy Noel Martha Dyer-Edwards)


In [68]:
df_titanic.loc[df_titanic['FirstName'] == ' Marguerite']

PassengerId  Survived Pclass                                Name     Sex  \
10           11         1  Third  Sandstrom, Miss. Marguerite Ru&5$$  female   

   Age  SibSp Parch   Ticket  Fare Cabin Embarked     ship    FirstName  \
10   4      1     1  PP 9549  16,7    G6        S  Titanic   Marguerite   

     LastName  Title  
10  Sandstrom   Miss

#### I think we have fixed the FirstName column. now, I will look at the LastName column

### LastName

In [69]:
for x in df_titanic['LastName']:
    for c in special_characters:
        if c in x:
            print(x)

Nicola-Yarred
Arnold-Franchi
Nicola-Yarred
Kink-Heilmann
Andersen-Jensen
Williams-Lambert
Arnold-Franchi
Bjornstrom-Steffansson
Johannesen-Bratthammer
Frolicher-Stehli
Stahelin-Maeglin
Simonius-Blumer


#### Seems like there is no improper values in LastName column.

In [70]:
df_titanic['LastName'].duplicated().sum()

223

In [71]:
df_titanic['LastName'].isnull().sum()

0

#### Seems like everything is okay with the LastName column.

### Sex

In [72]:
df_titanic['Sex'].isnull().sum()

0

In [73]:
df_titanic['Sex'].value_counts()

male       574
female     307
fem          2
malef        1
mal          1
femmale      1
feemale      1
Female       1
malee        1
F            1
Name: Sex, dtype: int64

#### Seems like we have some improper values in 'Sex' column.

#### I will fix these values with the correct ones: 'male' and 'female'.

In [74]:
df_titanic['Sex'] = df_titanic['Sex'].replace(['fem', 'femmale', 'feemale', 'Female', 'F'], 'female')

In [75]:
df_titanic['Sex'] = df_titanic['Sex'].replace(['malef', 'mal', 'malee'], 'male')

In [76]:
df_titanic['Sex'].value_counts()

male      577
female    313
Name: Sex, dtype: int64

In [77]:
df_titanic['Sex'].value_counts().sum()

890

#### Now, It's all good with the Sex column.

### Age

In [78]:
df_titanic['Age'].isnull().sum()

173

In [79]:
(df_titanic['Age'].isnull().sum() / df_titanic['Age'].value_counts().sum()) * 100

24.12831241283124

#### We have approx %24 null values so we have deal with these values. 

#### First we have to deal with the object values because we need to have float or integer values.

In [80]:
df_titanic['Age'].describe()

count     717
unique     94
top        24
freq       30
Name: Age, dtype: object

#### I think we have some string values with ',' so I will deal with them first.

In [81]:
for x in df_titanic['Age']:
    if type(x) == str:
        if ',' in x:
            print(x)

28,5
0,83
14,5
70,5
32,5
32,5
36,5
55,5
40,5
45,5
20,5
23,5
0,92
45,5
0,75
40,5
0,75
24,5
28,5
0,67
30,5
0,42
30,5
0,83
34,5


In [82]:
df_titanic['Age'] = df_titanic['Age'].str.replace(',', '.')

In [83]:
for x in df_titanic['Age']:
    if type(x) == str:
        if ',' in x:
            print(x)

#### Here, I have fixed them now I will look for other improper values and then will change them to integers.

In [84]:
df_titanic['Age'] = df_titanic['Age'].astype(float)

In [85]:
for x in df_titanic['Age']:
    if type(x) == str:
        print(x)

In [86]:
df_titanic['Age'].dtype

dtype('float64')

#### And I will fix the improper values

In [87]:
for x in df_titanic['Age']:
    if type(x) == float:
        if x > 102 or x < 0:
            print(x)

4435.0
-3.0
-12.0
250.0


#### We have 4 improper values because usually nobody can be older than 102 years old. And it is impossible to be younger than 0. Maybe I can fix the (-3,-12,250) values with (3,12,25) but I don't know what to do with 4435 so I will change it to wrong value.

In [88]:
df_titanic['Age'] = df_titanic['Age'].replace([-3.0,-12.0,250.0], [3.0,12.0,25.0])

In [89]:
df_titanic['Age'] = df_titanic['Age'].replace(4435.0, 'WrongValue')

In [90]:
for x in df_titanic['Age']:
    if type(x) == str:
        if x == 'WrongValue':
            print(x)

WrongValue


#### Now, we can fill in the null values

In [91]:
df_titanic['Age'].isnull().sum()

173

In [92]:
df_titanic['Age'] = df_titanic['Age'].fillna('NoInfo')

### SibSp

In [93]:
df_titanic['SibSp'].isnull().sum()

0

#### We don't have any null values in SibSp column. I will check for the improper values.

In [94]:
df_titanic['SibSp'].unique()

array([ 1,  0,  3,  4,  2,  5,  8, 70], dtype=int64)

#### We have 70 value that might be 7 and mistyped.

In [95]:
df_titanic.loc[df_titanic['SibSp'] == 70]

PassengerId  Survived Pclass               Name   Sex   Age  SibSp Parch  \
462          463         0  First  Gee, Mr. Arthur H  male  47.0     70     0   

     Ticket  Fare Cabin Embarked     ship  FirstName LastName Title  
462  111320  38,5   E63        S  Titanic   Arthur H      Gee    Mr

#### I will change this value to 7 then I think everything will be all right.

In [96]:
df_titanic['SibSp'] = df_titanic['SibSp'].replace(70, 7)

In [97]:
df_titanic['SibSp'].unique()

array([1, 0, 3, 4, 2, 5, 8, 7], dtype=int64)

#### Everything looks good with the SibSp column right now.

### Parch

In [98]:
df_titanic['Parch'].isnull().sum()

0

#### We don't have any null values in Parch column. I will check for the improper values.

In [99]:
df_titanic['Parch'].unique()

array(['0', '1', '2', '5', '3', '4', 'no'], dtype=object)

#### As I can see from here we have 2 problems. First one is data type and the second one is we have no value, we should change it to 0.

#### First, I will check for the data type for 0 value to make no as the same.

In [100]:
for x in df_titanic['Parch']:
    if x == '0':
        print(type(x))

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class

#### They are strings so we can change no to 0 string, then I will change the data type of this column.

In [101]:
df_titanic['Parch'] = df_titanic['Parch'].str.replace('no', '0')

In [102]:
df_titanic['Parch'].unique()

array(['0', '1', '2', '5', '3', '4'], dtype=object)

In [103]:
df_titanic['Parch'] = df_titanic['Parch'].astype(int)

In [104]:
df_titanic['Parch'].info()

<class 'pandas.core.series.Series'>
Int64Index: 890 entries, 0 to 893
Series name: Parch
Non-Null Count  Dtype
--------------  -----
890 non-null    int32
dtypes: int32(1)
memory usage: 10.4 KB


In [105]:
df_titanic['Parch'].unique()

array([0, 1, 2, 5, 3, 4])

#### Now, everything looks good with the Parch column.

### Ticket

In [106]:
df_titanic['Ticket'].isnull().sum()

1

In [107]:
df_titanic['Ticket'] = df_titanic['Ticket'].fillna('NoInfo')

In [108]:
df_titanic['Ticket'].isnull().sum()

0

#### I will check for the unique values in this column.

In [109]:
df_titanic['Ticket'].unique()

array(['A/5 21171', 'PC 17599', 'STON/O2. 3101282', '113803', '373450',
       '330877', '17463', '349909', '347742', '237736', 'PP 9549',
       '113783', 'A/5. 2151', '350406', '248706', '382652', '244373',
       '345763', '2649', '239865', '248698', '330923', '347077', '2631',
       '19950', '330959', '349216', 'PC 17601', 'PC 17569', '335677',
       'C.A. 24579', 'PC 17604', '113789', '2677', 'A./5. 2152', '345764',
       '2651', '7546', '11668', '349253', 'SC/Paris 2123', '330958',
       'S.C./A.4. 23567', '370371', '14311', '2662', '349237', '3101295',
       'A/4. 39886', 'PC 17572', '2926', '113509', '19947', 'C.A. 31026',
       '2697', 'C.A. 34651', 'CA 2144', '2669', '113572', '36973',
       '347088', 'PC 17605', '2661', 'C.A. 29395', 'S.P. 3464', '3101281',
       '315151', 'C.A. 33111', 'S.O.C. 14879', '2680', '1601', '348123',
       '349208', '374746', '248738', '364516', '345767', '345779',
       '330932', '113059', 'SO/C 14885', '3101278', 'W./C. 6608',
       '

#### We should have numbers only as I can see from here because we mostly have only numbers.

In [110]:
numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
for x in df_titanic['Ticket']:
    if type(x) == str:
        flag = True
        for y in numbers:
            if y in x:
                flag = False
                break
    
        if flag == True:
            print(x)

LINE
LINE
LINE
LINE
NoInfo


#### As can be seen from above we have some 'LINE' values. They don't have numbers inside so we have to change them to 'WrongValue'.

In [111]:
df_titanic['Ticket'] = df_titanic['Ticket'].str.replace('LINE', 'WrongValue')

In [112]:
numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
for x in df_titanic['Ticket']:
    if type(x) == str:
        flag = True
        for y in numbers:
            if y in x:
                flag = False
                break
    
        if flag == True:
            print(x)

WrongValue
WrongValue
WrongValue
WrongValue
NoInfo


#### Now, we can deal with the other values.

In [113]:
for x in df_titanic['Ticket']:
    if ' ' in x:
        print(x)

A/5 21171
PC 17599
STON/O2. 3101282
PP 9549
A/5. 2151
PC 17601
PC 17569
C.A. 24579
PC 17604
A./5. 2152
SC/Paris 2123
S.C./A.4. 23567
A/4. 39886
PC 17572
C.A. 31026
C.A. 34651
CA 2144
PC 17605
C.A. 29395
S.P. 3464
C.A. 33111
CA 2144
S.O.C. 14879
SO/C 14885
W./C. 6608
SOTON/OQ 392086
W.E.P. 5734
C.A. 2315
PC 17754
PC 17759
STON/O 2. 3101294
PC 17558
S.O.C. 14879
A4. 54510
C 17369
SOTON/O.Q. 3101307
A/5. 3337
C.A. 29178
SC/PARIS 2133
PC 17593
STON/O2. 3101279
C.A. 33112
W./C. 6608
S.O.P. 1166
A.5. 11206
A/5. 851
Fa 265302
PC 17597
SOTON/OQ 392090
CA. 2343
C.A. 33595
PC 17318
STON/O 2. 3101280
PC 17595
CA. 2343
SC/PARIS 2131
PC 17610
PC 17569
CA. 2343
A/5 3540
SOTON/O.Q. 3101311
F.C.C. 13528
A/5 21174
STON/O2. 3101283
W/C 14208
SOTON/OQ 392089
PP 4348
SW/PP 751
A/5 21173
C.A. 29566
W./C. 6609
C.A. 31921
SCO/W 1585
W./C. 14263
STON/O 2. 3101275
A/5. 3336
PC 17585
PC 17755
C.A. 17248
PC 17582
PC 17760
PC 17596
C.A. 2673
A/5. 10482
SC/Paris 2163
PC 17612
PC 17558
A/5 2466
PC 17758
P/PP 3381
P

#### Maybe we can delete the codes in the first part, but instead I will leave them like this for now. And I will check for the duplicate values.

In [114]:
df_titanic['Ticket'].duplicated().sum()

209

#### I don't know if they have the same ticket as can be seen from the amount they can so I think Ticket column is ok.

### Fare

In [115]:
df_titanic['Fare'].isnull().sum()

1

#### We have 1 null value in the Fare column. I will now check for other problems.

In [116]:
df_titanic['Fare'].value_counts()

8,05       43
13         42
7,8958     38
7,75       34
26         31
           ..
35          1
15,9        1
28,5        1
6,2375      1
10,5167     1
Name: Fare, Length: 251, dtype: int64

#### At the first eye we have problems with the float typings.

In [117]:
df_titanic['Fare'].unique()

array(['7,25', '71,2833', '7,925', '53,1', '8,05', '8,4583', '51,8625',
       '21,075', '11,1333', '30,0708', '16,7', '26,55', '7,8542', '16',
       '29,125', '13', '18', '7,225', '26', '8,0292', '31,3875', '263',
       '7,8792', '7,8958', '27,7208', '146,5208', '7,75', '10,5',
       '82,1708', '52', '7,2292', '11,2417', '9,475', '21', '41,5792',
       '15,5', '21,6792', '17,8', '39,6875', '7,8', '76,7292', '61,9792',
       '35,5', '27,75', '46,9', '80', '83,475', '27,9', '15,2458',
       '8,1583', '8,6625', '73,5', '14,4542', '56,4958', '7,65', '29',
       '12,475', '9', '9,5', '7,7875', '47,1', '15,85', '34,375',
       '61,175', '20,575', '34,6542', '63,3583', '23', '77,2875',
       '8,6542', '7,775', '24,15', '9,825', '14,4583', '247,5208',
       '31,275', '7,1417', '22,3583', '6,975', '7,05', '14,5', '15,0458',
       '26,2833', '9,2167', '79,2', '6,75', '11,5', '36,75', '7,7958',
       '12,525', '66,6', '7,3125', '61,3792', '7,7333', '69,55', '16,1',
       '15,75', '2

#### I will try to use some scripts to search for other problems. For exp. values below 0.

#### But first we have to fix data type problem to find all values that are below 0.

In [118]:
for x in df_titanic['Fare']:
    if type(x) == str:
        if ',' in x:
            print(x)

7,25
71,2833
7,925
53,1
8,05
8,4583
51,8625
21,075
11,1333
30,0708
16,7
26,55
8,05
7,8542
29,125
7,225
8,0292
21,075
31,3875
7,225
7,8792
7,8958
27,7208
146,5208
7,75
10,5
82,1708
7,2292
8,05
11,2417
9,475
7,8958
41,5792
7,8792
8,05
15,5
7,75
21,6792
17,8
39,6875
7,8
76,7292
61,9792
35,5
10,5
7,2292
27,75
46,9
7,2292
83,475
27,9
27,7208
15,2458
10,5
8,1583
7,925
8,6625
10,5
46,9
73,5
14,4542
56,4958
7,65
7,8958
8,05
12,475
9,5
7,7875
47,1
10,5
15,85
34,375
8,05
8,05
8,05
7,8542
61,175
20,575
7,25
8,05
34,6542
63,3583
7,8958
7,8958
77,2875
8,6542
7,925
7,8958
7,65
7,775
7,8958
24,15
14,4542
8,05
9,825
14,4583
7,925
7,75
247,5208
31,275
73,5
8,05
30,0708
77,2875
11,2417
7,75
7,1417
22,3583
6,975
7,8958
7,05
14,5
15,0458
26,2833
53,1
9,2167
79,2
15,2458
7,75
15,85
6,75
11,5
36,75
7,7958
34,375
12,525
66,6
8,05
14,5
7,3125
61,3792
7,7333
8,05
8,6625
69,55
16,1
15,75
7,775
8,6625
39,6875
20,525
27,9
25,925
56,4958
33,5
29,125
11,1333
7,925
30,6958
7,8542
25,4667
28,7125
69,55
15,05
31,3875


In [119]:
df_titanic['Fare'] = df_titanic['Fare'].str.replace(',', '.')

In [120]:
for x in df_titanic['Fare']:
    if type(x) == str:
        if ',' in x:
            print(x)

#### Now, I will change the data type.

In [121]:
#df_titanic['Fare'] = df_titanic['Fare'].astype(float)

#### I have tried to run this code but it gave an error like you have a '07.maj' in the column. So I will take care of this problem.

In [122]:
df_titanic['Fare'] = df_titanic['Fare'].str.replace('07.maj', '7')

C:\Users\emir1\AppData\Local\Temp\ipykernel_25404\3334570278.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df_titanic['Fare'] = df_titanic['Fare'].str.replace('07.maj', '7')


In [123]:
df_titanic['Fare'] = df_titanic['Fare'].astype(float)

#### Now we can search for the other improper values.

In [124]:
for x in df_titanic['Fare']:
    if x < 0:
        print(x)

-90.0


#### We have a minus value I will fix the sign.

In [125]:
df_titanic['Fare'] = df_titanic['Fare'].replace(-90.0, 90.0)

#### We may have some so high values but I don't know if they are correct so I will leave them like this but if it is we can normalize them in the future.

### Cabin

In [126]:
df_titanic['Cabin'].isnull().sum()

686

#### And now there is a big problem. I don't know if I should drop this column or not but I will fill the null values with something for now. But we can drop it in the future.

In [127]:
df_titanic['Cabin'] = df_titanic['Cabin'].fillna('NoInfo')

In [128]:
df_titanic['Cabin'].isnull().sum()

0

#### And I will look for the improper values right now.

In [129]:
df_titanic['Cabin'].value_counts()

NoInfo         686
B96 B98          4
G6               4
C93              4
C23 C25 C27      4
              ... 
C32              1
E34              1
C7               1
C54              1
C148             1
Name: Cabin, Length: 146, dtype: int64

In [130]:
df_titanic['Cabin'].unique()

array(['NoInfo', 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'C23 C25 C27',
       'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33', 'F G73', 'E31',
       'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101', 'F E69', 'D47',
       'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4', 'A32', 'B4',
       'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35', 'C87', 'B77',
       'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19', 'B49', 'D',
       'C22 C26', 'C106', 'C65', 'E36', 'C54', 'B57 B59 B63 B66', 'C7',
       'E34', 'C32', 'B18', 'C124', 'C91', 'E40', 'T', 'C128', 'D37',
       'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44', 'A34', 'C104',
       'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14', 'B37', 'C30',
       'D20', 'B79', 'E25', 'D46', 'C95', 'B38', 'B39', 'B22', 'C86',
       'C70', 'A16', 'C101', 'C68', 'A10', 'E68', 'B41', 'A20', 'D19',
       'D50', 'D9', 'A23', 'B50', 'A26', 'D48', 'E58', 'C126', 'B71',
       'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63', 'C62 C64', 'E24',
       

#### At the first eye I can see that we have so many improper values in this column. So I will drop this column.

In [131]:
df_titanic.drop('Cabin', axis=1, inplace=True)

In [132]:
df_titanic.head(5)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked     ship  \
0      0         A/5 21171   7.2500        S  Titanic   
1      0          PC 17599  71.2833        C  Titanic   
2      0  STON/O2. 3101282   7.9250        S  Titanic   
3      0            113803  53.1000        S  Titanic   
4      0            373450   8.0500        S  Titanic   

                                FirstName   LastName  Title  
0                             Owen Harris     Braund     Mr  
1   John Bradley (Florence Briggs Thayer)    Cumings    Mrs  
2                                   Laina  Heikkinen   Miss  
3           Jacques Heath (Lily May Peel)   Futrelle    Mrs  
4                           William Henry      Allen     Mr

### Embarked

In [133]:
df_titanic['Embarked'].isnull().sum()

2

In [134]:
df_titanic['Embarked'] = df_titanic['Embarked'].fillna('NoInfo')

#### Filled the null values.

In [135]:
df_titanic['Embarked'].unique()

array(['S', 'C', 'Q', 'So', 'NoInfo', 'Co', 'Qe'], dtype=object)

#### This column has some categorized scheme but have some improper inputs in the column I will fix them.

In [136]:
df_titanic['Embarked'] = df_titanic['Embarked'].replace(['So', 'Co', 'Qe'], ['S', 'C', 'Q'])

In [137]:
df_titanic['Embarked'].unique()

array(['S', 'C', 'Q', 'NoInfo'], dtype=object)

In [138]:
df_titanic['Embarked'].count()

890

### Ship

In [139]:
df_titanic['ship'].isnull().sum()

0

#### We don't have any null values, I will check for improper values.

In [140]:
df_titanic['ship'].unique()

array(['Titanic', 'Titani', 'Titnic'], dtype=object)

In [141]:
df_titanic['ship'] = df_titanic['ship'].replace(['Titani', 'Titnic'], 'Titanic')

In [142]:
df_titanic['ship'].unique()

array(['Titanic'], dtype=object)

### Final Look

In [143]:
df_titanic.head(5)

PassengerId  Survived Pclass  \
0            1         0  Third   
1            2         1  First   
2            3         1  Third   
3            4         1  First   
4            5         0  Third   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked     ship  \
0      0         A/5 21171   7.2500        S  Titanic   
1      0          PC 17599  71.2833        C  Titanic   
2      0  STON/O2. 3101282   7.9250        S  Titanic   
3      0            113803  53.1000        S  Titanic   
4      0            373450   8.0500        S  Titanic   

                                FirstName   LastName  Title  
0                             Owen Harris     Braund     Mr  
1   John Bradley (Florence Briggs Thayer)    Cumings    Mrs  
2                                   Laina  Heikkinen   Miss  
3           Jacques Heath (Lily May Peel)   Futrelle    Mrs  
4                           William Henry      Allen     Mr

In [144]:
df_titanic.drop('Name', axis=1, inplace=True)

In [145]:
df_titanic = df_titanic[['PassengerId', 'Survived', 'Pclass', 'Title', 'FirstName', 'LastName', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Embarked', 'ship']]

In [147]:
df_titanic.head(5)

PassengerId  Survived Pclass  Title  \
0            1         0  Third     Mr   
1            2         1  First    Mrs   
2            3         1  Third   Miss   
3            4         1  First    Mrs   
4            5         0  Third     Mr   

                                FirstName   LastName     Sex   Age  SibSp  \
0                             Owen Harris     Braund    male  22.0      1   
1   John Bradley (Florence Briggs Thayer)    Cumings  female  38.0      1   
2                                   Laina  Heikkinen  female  26.0      0   
3           Jacques Heath (Lily May Peel)   Futrelle  female  35.0      1   
4                           William Henry      Allen    male  35.0      0   

   Parch            Ticket     Fare Embarked     ship  
0      0         A/5 21171   7.2500        S  Titanic  
1      0          PC 17599  71.2833        C  Titanic  
2      0  STON/O2. 3101282   7.9250        S  Titanic  
3      0            113803  53.1000        S  Titanic  
4      0            373450   8.0500        S  Titanic

### Writing to a new tsv file.

In [148]:
df_titanic.to_csv('NewTitanic.tsv', sep='\t')